src/TokenHandler.mo

```diff
@@ -0,0 +1,364 @@
+        tree.entries(),
+        func((principal, info)) = info.credit != 0 or info.deposit != 0,
+      )
+    );
+
+    /// deserialize tracking data
+    public func unshare(values : [(Principal, Info)]) {
+      tree := RBTree.RBTree<Principal, InfoLock>(Principal.compare);
+      for ((principal, value) in values.vals()) {
+        tree.put(
+          principal,
+          {
+            var deposit = value.deposit;
+            var credit = value.credit;
+            var lock = false;
+          },
+        );
+      };
+    };
+  };
+
+  public class BackLog() {
+    // a backlog of principals, waiting for consolidation
+    var backlog : AssocList.AssocList<Principal, ()> = null;
+    var size_ : Nat = 0;
+
+    public func push(principal : Principal) {
+      let (updated, prev) = AssocList.replace<Principal, ()>(backlog, principal, Principal.equal, ?());
+      backlog := updated;
+      switch (prev) {
+        case (null) size_ += 1;
+        case (_) {};
+      };
+    };
+
+    public func remove(principal : Principal) {
+      let (updated, prev) = AssocList.replace<Principal, ()>(backlog, principal, Principal.equal, null);
+      backlog := updated;
+      switch (prev) {
+        case (null) {};
+        case (_) size_ -= 1;
+      };
+    };
+
+    /// retrieve the current size of consolidation backlog
+    public func size() : Nat = size_;
+
+    public func pop() : ?Principal {
+      switch (backlog) {
+        case (null) null;
+        case (?((principal, _), list)) {
+          backlog := list;
+          size_ -= 1;
+          ?principal;
+        };
+      };
+    };
+
+    public func share() : [Principal] {
+      func fold(list : AssocList.AssocList<Principal, ()>) : Buffer.Buffer<Principal> {
+        let ?((principal, _), tail) = list else return Buffer.Buffer<Principal>(0);
+        let a = fold(tail);
+        a.add(principal);
+        a;
+      };
+
+      Buffer.toArray(fold(backlog));
+    };
+
+    public func unshare(array : [Principal]) {
+      backlog := null;
+      size_ := array.size();
+      var i = size_;
+      while (i > 0) {
+        backlog := ?((array[i - 1], ()), backlog);
+        i -= 1;
+      };
+    };
+  };
+
+  public class TokenHandler(
+    icrc1LedgerPrincipal : Principal,
+    ownPrincipal : Principal,
+    fee : Nat,
+  ) {
+
+    let icrc1Ledger = actor (Principal.toText(icrc1LedgerPrincipal)) : ICRC1Interface.Icrc1LedgerInterface;
+
+    // The map from principal to tracking info:
+    // var tree : RBTree.RBTree<Principal, InfoLock> = RBTree.RBTree<Principal, InfoLock>(Principal.compare);
+
+    let map : Map = Map();
+    let backlog : BackLog = BackLog();
+
+    // a backlog of principals, waiting for consolidation
+    // var backlog : AssocList.AssocList<Principal, ()> = null;
+    // var backlogSize : Nat = 0;
+
+    /// query the usable balance
+    public func balance(principal : Principal) : Nat = info(principal).usable_balance;
```

Comment:
Let's use `p : Principal` everywhere.

Philosophy: use verbose names if they add information. But using the type name as variable name isn't additional information. For example, we wouldn't write `var nat : Nat`.